pkg/servicemesh/controller/memberroll/controller.go

```diff
@@ -121,56 +142,53 @@ func (smmrc *serviceMeshMemberRollController) getNamespaces(namespaces []string)
+func (smmrc *serviceMeshMemberRollController) getSeedNamespaces() []string {
+	smmrc.lock.RLock()
+	defer smmrc.lock.RUnlock()
+	return smmrc.seedNamespaces
+}
+
+func (smmrc *serviceMeshMemberRollController) setSeedNamespaces(seedNamespaces []string) {
+	sort.Strings(seedNamespaces)
+	smmrc.lock.Lock()
+	defer smmrc.lock.Unlock()
+	smmrc.seedNamespaces = seedNamespaces
+}
+
+func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(listener Listener, name string) cache.ResourceEventHandler {
 	handler := &serviceMeshMemberRollListener{
 		smmrc:             smmrc,
 		listener:          listener,
 		currentNamespaces: nil,
 		name:              name,
-		seedCh:            make(chan struct{}),
-	}
-
-	// Previously we sent an immediate initial update to all listeners when they
-	// were registered that included only the Istio system namespace.  That
-	// caused problems with some controllers, e.g. IOR, because they expect the
-	// callback to be called with the full authoritative set of namespaces and
-	// may remove resources for namespaces not in the list.
-	//
-	// This instead waits for the informer's cache to sync, then sends an
-	// initial update only if the expected SMMR is not found in the cache.
-	go func() {
-		_ = wait.PollImmediateInfinite(100*time.Millisecond, func() (done bool, err error) {
-			smmrc.cacheLock.RLock()
-			defer smmrc.cacheLock.RUnlock()
-
-			return smmrc.cacheWarmed, nil
-		})
-
-		smmrLog.Infof("Cache synced for listener %q", name)
-
-		// Closing seedCh allows the handler to start processing events.
-		defer close(handler.seedCh)
-
-		cacheKey := smmrc.namespace + "/" + smmrc.memberRollName
-		_, exists, _ := smmrc.informer.GetStore().GetByKey(cacheKey)
-		if exists {
-			// No need to send initial update.  The informer will do it.
-			return
-		}
+	}
+	handler.updateNamespaces("add", smmrc.memberRollName, smmrc.getNamespaces(smmrc.getSeedNamespaces()))
+	return handler
+}
+
+func (smmrc *serviceMeshMemberRollController) OnAdd(obj interface{}) {
+	if serviceMeshMemberRoll := getServiceMeshMemberRoll(obj); serviceMeshMemberRoll != nil {
+		smmrc.setSeedNamespaces(serviceMeshMemberRoll.Status.ConfiguredMembers)
+	}
+}
 
-		smmrLog.Infof("Seeding listener %q with system namespace.", name)
-		handler.updateNamespaces("seed", smmrc.memberRollName, nil)
-	}()
+func (smmrc *serviceMeshMemberRollController) OnUpdate(oldObj, newObj interface{}) {
+	if serviceMeshMemberRoll := getServiceMeshMemberRoll(newObj); serviceMeshMemberRoll != nil {
+		smmrc.setSeedNamespaces(serviceMeshMemberRoll.Status.ConfiguredMembers)
+	}
+}
 
-	return handler
+func (smmrc *serviceMeshMemberRollController) OnDelete(obj interface{}) {
+	if serviceMeshMemberRoll := getServiceMeshMemberRoll(obj); serviceMeshMemberRoll != nil {
+		smmrc.setSeedNamespaces(serviceMeshMemberRoll.Status.ConfiguredMembers)
```

Comment:
Should it be set to nil regardless?

Reply:
If there are no configured namespaces then the update will see only the control plane namespace which would be appropriate, as other namespaces are created/reconciled they will receive further updates.

Reply:
Ah sorry, I follow what you are saying and it should be nil

Reply:
done